src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs

```diff
@@ -388,47 +397,78 @@ private static bool IsRoutePartNonEmpty(object routePart)
         private struct TemplateBindingContext
         {
             private readonly RouteValueDictionary _defaults;
-            private readonly RouteValueDictionary _acceptedValues;
+            private RouteValueDictionary _acceptedValues;
+            private readonly RouteValueDictionary _values;
 
-            public TemplateBindingContext(RouteValueDictionary defaults)
+            public TemplateBindingContext(RouteValueDictionary defaults, RouteValueDictionary values)
             {
                 _defaults = defaults;
+                _values = values;
+                // Perf: In most of the common cases, AcceptedValues is same as the values.
+                // But when they are different, values is readonly and hence we need to create a new
+                // RouteValueDictionary for AcceptedValues.
+                // However to optimize the common cases, we initially start with null acceptedValues.
+                // We will initialize it only, when we find that values is different from acceptedValues.
+                // One case where we know that acceptedValues will be different from values is, when there is
+                // an empty value in values dictionary. If that is the case, we right away intialize the acceptedValues to 
+                // a new RouteValueDictionary. 
+                // This optimization saves one RouteValueDictionary allocation per GetValues call in common cases.
```

Comment:
As per my understanding, for all the paths where the method GetValues is called internally in MVC doesn't have this case of empty value. But as this is open for extensibility and can be called outside of the internal code, this check may be necessary